metric/unit/unit.go

```diff
@@ -14,12 +14,43 @@
 
 package unit // import "go.opentelemetry.io/otel/metric/unit"
 
+// Dimensionless is the UCUM dimensionless unit 1.
+var Dimensionless = Unit{code: "1"}
+
+// prefix is a numerical modifier of a unit quantity. Its value is the
+// case-sensitive UCUM prefix code.
+type prefix string
+
 // Unit is a determinate standard quantity of measurement.
-type Unit string
-
-// Units defined by OpenTelemetry.
-const (
-	Dimensionless Unit = "1"
-	Bytes         Unit = "By"
-	Milliseconds  Unit = "ms"
-)
+type Unit struct {
+	prefix prefix
+
+	code string
+}
+
+// New returns a new Unit representing the unit code. The code should be the
+// UCUM (https://ucum.org/ucum) case-sensitive code as this is what
+// OpenTelemetry uses.
+func New(code string) Unit { return Unit{code: code} }
+
+// withPrefix returns a copy of u with prefix p.
+func (u Unit) withPrefix(p prefix) Unit {
+	u.prefix = p
```

Comment:
I think that we should not allow setting the `prefix` twice

```suggestion
	if u.prefix != "" {
		panic("unit prefix cannot be overridden")
	}
	u.prefix = p
```

Reply:
Given this copies the unit, I think having the copy of the unit have the new prefix and the original unaffected seems like the appropriate behavior here.

Reply:
If I follow correctly then `unit.Nano(unit.Milli(unit.Second))` would result in having `1ns`. I was thinking about panicking in such case.

Reply:
After second thought I see a drawback in my proposal which is performance overhead. Probably it does not make sense to be "over-protective" here.

Reply:
> If I follow correctly then `unit.Nano(unit.Milli(unit.Second))` would result in having `1ns`. I was thinking about panicking in such case.

Correct in the end you would have a unit that represented `1ns`. I think this behavior is clearly communicated to the user in the `Nano` and `Milli` documentation that they will each set the prefix for the returned copy of the unit. If a user wants to run `unit.Nano(unit.Milli(unit.Second))` and expect a different outcome I'm not going to be able to save them from themselves here.